src/core/io/platform/f4xx/ADCf4xx.cpp

```diff
@@ -1,64 +1,104 @@
- * DISCLAIMER: THIS MIGHT BREAK AT ANY POINT AND/OR NOT WORK FOR CERTAIN PURPOSES
- *
+/*
  * This DMA ADC is different than f3xx DMA ADC!
- * This DMA ADC does NOT use the interrupts, as when interrupts were enabled it
- * would constantly interrupt, not letting the print statements or anything
- * else happen.
- * Tried to fix this by changing the sampletime and clock prescaler values, but
- * nothing worked.
- * This was "fixed" by commenting out the NVIC_EnableIRQ, stopping the interrupt
- * from ever being enabled.
+ * f4xx DMA ADC uses a timer to trigger ADC conversions. Timer frequency is 1kHz.
  *
- * For commit from before code was removed, refer to commit 81624521a8b2c4b66480193e88cf32782aaee84d.
+ * Timers were added to slow down ADC DMA interrupts, as when they ADC DMA is allowed to convert constantly,
+ * the interrupts stopped the program from doing anything else besides interrupt calls.
+ *
+ * WARNING: DOES NOT WORK ON PINS THAT ARE ALREADY IN USE (LIKE UART PINS)
  */
 
 #include <HALf4/stm32f4xx.h>
 #include <HALf4/stm32f4xx_hal_adc.h>
 #include <core/io/pin.hpp>
 #include <core/io/platform/f4xx/ADCf4xx.hpp>
 #include <core/io/platform/f4xx/GPIOf4xx.hpp>
+#include <core/platform/f4xx/stm32f4xx.hpp>
+#include <core/utils/log.hpp>
+
+namespace {
+/// This is made as a global variable so that it is accessible in the interrupt.
+DMA_HandleTypeDef* dmaHandle[3];
+ADC_HandleTypeDef* adcHandle[3];
+
+/**
+ * This function handles DMA2 stream0 global interrupt. (For ADC 1)
+ */
+extern "C" void DMA2_Stream0_IRQHandler(void) {
+    HAL_DMA_IRQHandler(dmaHandle[0]);
+    HAL_ADC_IRQHandler(adcHandle[0]);
+}
+
+/**
+ * This function handles DMA2 stream2 global interrupt. (For ADC 2)
+ */
+extern "C" void DMA2_Stream2_IRQHandler(void) {
+    HAL_DMA_IRQHandler(dmaHandle[1]);
+    HAL_ADC_IRQHandler(adcHandle[1]);
+}
+
+/**
+ * This function handles DMA2 stream1 global interrupt. (For ADC 3)
+ * This is correct, stream 1 for ADC 3. STM is odd and makes things confusing
+ */
+extern "C" void DMA2_Stream1_IRQHandler(void) {
+    HAL_DMA_IRQHandler(dmaHandle[2]);
+    HAL_ADC_IRQHandler(adcHandle[2]);
+}
+} // namespace
 
 namespace core::io {
+constexpr uint8_t ADC1_SLOT = 0;
+constexpr uint8_t ADC2_SLOT = 1;
+constexpr uint8_t ADC3_SLOT = 2;
 
-// Init static member variables
-ADC_HandleTypeDef ADCf4xx::halADC = {0};
-Pin ADCf4xx::channels[MAX_CHANNELS];
-uint16_t ADCf4xx::buffer[MAX_CHANNELS];
-DMA_HandleTypeDef ADCf4xx::halDMA = {0};
+constexpr uint8_t ADC1SHIFT = 5;
+constexpr uint8_t ADC2SHIFT = 6;
+constexpr uint8_t ADC3SHIFT = 7;
 
-ADCf4xx::ADCf4xx(Pin pin) : ADC(pin) {
-    // Flag representing if the ADC has been configured yet
-    static bool halADCisInit = false;
+// Combines the channel memory value with the ADC peripherals it supports into one uint32_t
+constexpr uint32_t CHANNEL_SET(uint8_t adc1, uint8_t adc2, uint8_t adc3, uint32_t ch) {
+    return (ch | (adc1 << ADC1SHIFT) | (adc2 << ADC2SHIFT) | (adc3 << ADC3SHIFT));
+}
```

Comment:
@mjh9585 How do we like replacing macros with `constexpr` functions? It's something I learned on co-op, and I really like using them

Reply:
I talked to Helldog for a while, and we came to the conclusion that `constexpr` functions can potentially problematic, but they're something we should look into in the future. For now, we agreed that it would be better to get rid of this entirely and use a packed bit field struct. This will have the memory savings of packing the bits ourselves, and it also allows us to get rid of the ugly bit masking stuff

Reply:
Make sure you also update F3 to be the same

Reply:
I'll be honest, ion know how packed bit structs are done in c++. I know its just like using all the bits in a struct so that there isn't empty/wasted bits inbetween variables. I also know there is an impact on the order of the variables you put into the struct but thats all i got. I'll google n look into packed bit structs and how to make/define them in c++. 

Just to clarify, I am creating a struct for the channel uint and the adc support bits, and then have that struct be created instead of calling this CHANNEL_SET() constexpr? 
What about the ADC1_SLOT and the other slots, which are also constexprs. Should I change them back into define macros, so we arent using any constexpr?

Reply:
The other question I have is, with the packed bit struct, should I have the channel be only 5 bits, because the largest the channel ever is is only 5 bits, or should I give it the full value that it starts as (uint32_t)

Reply:
The `constexpr` variables are good as they are. The concern is the `constexpr` macro.

You should be able to put the packed struct into a single byte, and that would be ideal in terms of memory, so you should try to do that